### PR TITLE
Add list collector summary block type

### PR DESCRIPTION
**Pull request by berroar**
### What is the context of this PR?
There was a requirement to hide a label being displayed on list collector summary pages [(see Trello)](https://trello.com/c/6V09Qp8L/3311-list-collector-summary-shouldnt-show-section-complete-panel-m). 

In order to make this change I added a new type of summary page: `ListCollectorSummary`. I have also done some further refactoring to the view contexts and block handlers so that they use a more consistent approach.

### How to review
Use the `test_list_collector_section_summary` schema to check that the list collector summary page is displaying the correct information. The` Check your answers before continuing to the next section` label should no longer be displayed.

Check that the same changes have been implemented in the census schemas.

### CODE REVIEW

File: scripts/test_schemas.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ if [ "$1" == "--local" ] || [ "$2" == "--local" ]; then
 fi
 
 if [ "$run_docker" == true ]; then
-    branch=v3
+    branch=add_list_section_summary_type
```
Comment (berroar) [inline scripts/test_schemas.sh]: Revert before merge

**Comment by berroar** [comment]
[list collector summary shouldnt show section complete panel (M)](https://trello.com/c/6V09Qp8L/3311-list-collector-summary-shouldnt-show-section-complete-panel-m)

File: tests/functional/generate_pages.py [hunk 2]
```diff
@@ -382,13 +382,31 @@ def process_calculated_summary(answers, page_spec):
...
+                LIST_SECTION_SUMMARY_REMOVE_LINK_GETTER.substitute(list_context)
+            )
+            page_spec.write(LIST_SECTION_SUMMARY_LABEL_GETTER.substitute(list_context))
+
+
+def process_summary(schema_data, page_spec):
+    for section in schema_data["sections"]:
         for group in section["groups"]:
             for block in group["blocks"]:
+                if block["type"] == "SectionSummary":
```
Comment (petechd) [inline tests/functional/generate_pages.py]: was moving this 'if' statement above 'for' loop necessary?

**Comment by iwootten** [comment]
I have previously wanted to [refactor the *summary* context files](https://github.com/ONSdigital/eq-survey-runner/pull/2313/files).

File: app/views/contexts/list_collector_context.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        return rendered_summary["item_title"]
+
+    def build_list_items_summary_context(self, list_collector_block, return_to=None):
+        list_name = list_collector_block["for_list"]
+        list_item_ids = self._list_store[list_name].items
+
+        primary_person = self._list_store[list_name].primary_person
+
+        placeholder_renderer = PlaceholderRenderer(
```
Comment (iwootten) [inline app/views/contexts/list_collector_context.py]: You can just use self._placeholder_renderer here.

Comment (berroar) [reply app/views/contexts/list_collector_context.py]: Fixed this.

File: app/views/contexts/list_collector_context.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from flask import url_for
+from flask_babel import lazy_gettext
+
+from app.views.contexts.context import Context
+from app.views.contexts.question import build_question_context
+
+
+class ListCollectorContext(Context):
+    @staticmethod
+    def get_item_title(
```
Comment (iwootten) [inline app/views/contexts/list_collector_context.py]: If this method was made static, placeholder_renderer could be removed as an argument here and the class renderer be used instead.

File: app/views/contexts/calculated_summary_context.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    get_answer_ids_in_block,
+)
+from app.views.contexts.context import Context
+from app.views.contexts.summary.group import Group
+
+
+class CalculatedSummaryContext(Context):
+    def build_groups_for_section(self, section):
+        router = Router(
+            self._schema,
```
Comment (iwootten) [inline app/views/contexts/calculated_summary_context.py]: I think you can use the base class router here.

File: app/views/contexts/list_collector_context.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                    )
+
+                if "remove_block" in list_collector_block:
+                    list_item_context["remove_link"] = url_for(
+                        "questionnaire.block",
+                        list_name=list_name,
+                        block_id=list_collector_block["remove_block"]["id"],
+                        list_item_id=list_item_id,
+                        return_to=return_to,
+                    )
```
Comment (JamesGardiner) [inline app/views/contexts/list_collector_context.py]: These could be partials, or you could instantiate the block_id in the if statements then call the url_for function outside of them

File: app/views/contexts/list_collector_summary_context.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+from flask import url_for
+
+from app.questionnaire.questionnaire_schema import QuestionnaireSchema
+from app.views.contexts.list_collector_context import ListCollectorContext
+from app.views.contexts.context import Context
+
+
+class ListCollectorSummaryContext(Context):
+    def build_view_context(self, current_location):
+
```
Comment (JamesGardiner) [inline app/views/contexts/list_collector_summary_context.py]:
```suggestion
```

File: tests/functional/generate_pages.py [hunk 8]
```diff
@@ -382,11 +382,27 @@ def process_calculated_summary(answers, page_spec):
         page_spec.write(CALCULATED_SUMMARY_LABEL_GETTER.substitute(answer_context))
 
 
-def process_summary(schema_data, page_spec):
+def process_list_collector_summary(schema_data, page_spec):
     for section in schema_data["sections"]:
         list_collector_blocks = QuestionnaireSchema.get_visible_list_blocks_for_section(
             section
         )
+        for list_instance, list_block in enumerate(list_collector_blocks):
```
Comment (JamesGardiner) [inline tests/functional/generate_pages.py]: This doesn't need to be enumerated as `list_instance` isn't used anywhere

File: .gitignore [hunk 9]
```diff
@@ -106,3 +106,4 @@ templates/layout
 *.mo
 
 .vim/
+data/
```
Comment (iwootten) [inline .gitignore]: This isn't needed anymore

File: app/questionnaire/router.py [hunk 10]
```diff
@@ -286,7 +286,7 @@ def _does_section_only_contain_summary(self, section_id):
     def _get_location_of_section_summary(self, routing_path):
         for location in routing_path[::-1]:
             block = self._schema.get_block(location.block_id)
-            if block["type"] == "SectionSummary":
+            if block["type"] in ["SectionSummary", "ListCollectorSummary"]:
```
[the remark this answers is not in the document]
Comment (ajmaddaford) [reply app/questionnaire/router.py]: @JamesGardiner that's not the same logic - currently it returns location if the block type matches, or moves on to the next block in the routing path.

[the remark this answers is not in the document]
Comment (ajmaddaford) [reply app/questionnaire/router.py]: @JamesGardiner that's not the same logic - currently it returns location if the block type matches, or moves on to the next block in the routing path.

**Pull request merged**